src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -1,30 +1,164 @@
 {
     public readonly partial struct DateTime
     {
-        internal static readonly bool s_systemSupportsLeapSeconds = SystemSupportsLeapSeconds();
+        private static LeapSecondCache? s_leapSecondCache;
+
+        private static unsafe delegate* unmanaged[Stdcall]<ulong*, void> s_pfnGetSystemTimeAsFileTime = GetGetSystemTimeAsFileTimeFnPtr();
+
+        private static unsafe delegate* unmanaged[Stdcall]<ulong*, void> GetGetSystemTimeAsFileTimeFnPtr()
+        {
+            IntPtr kernel32Lib = NativeLibrary.Load("kernel32.dll", typeof(DateTime).Assembly, DllImportSearchPath.System32);
+            IntPtr pfnGetSystemTime = NativeLibrary.GetExport(kernel32Lib, "GetSystemTimeAsFileTime"); // will never fail
+
+            if (NativeLibrary.TryGetExport(kernel32Lib, "GetSystemTimePreciseAsFileTime", out IntPtr pfnGetSystemTimePrecise))
+            {
+                // If GetSystemTimePreciseAsFileTime exists, we'd like to use it.  However, on
+                // misconfigured systems, it's possible for the "precise" time to be inaccurate:
+                //     https://github.com/dotnet/runtime/issues/9014
+                // If it's inaccurate, though, we expect it to be wildly inaccurate, so as a
+                // workaround/heuristic, we get both the "normal" and "precise" times, and as
+                // long as they're close, we use the precise one. This workaround can be removed
+                // when we better understand what's causing the drift and the issue is no longer
+                // a problem or can be better worked around on all targeted OSes.
+
+                // TODO: Once https://github.com/dotnet/runtime/issues/38134 is resolved,
+                // the calls below should suppress the GC transition.
+
+                ulong filetimeStd, filetimePrecise;
+                ((delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTime)(&filetimeStd);
+                ((delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTimePrecise)(&filetimePrecise);
+
+                if (Math.Abs((long)(filetimeStd - filetimePrecise)) <= 100 * TicksPerMillisecond)
+                {
+                    pfnGetSystemTime = pfnGetSystemTimePrecise; // use the precise version
+                }
+            }
+
+            return (delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTime;
+        }
 
         public static unsafe DateTime UtcNow
         {
             get
             {
-                if (s_systemSupportsLeapSeconds)
+                // The OS tick count and .NET's tick count are slightly different. The OS tick
+                // count is the *absolute* number of 100-ns intervals which have elapsed since
+                // January 1, 1601 (UTC). Due to leap second handling, the number of ticks per
+                // day is variable. Dec. 30, 2016 had 864,000,000,000 ticks (a standard 24-hour
+                // day), but Dec. 31, 2016 had 864,010,000,000 ticks due to leap second insertion.
```

Comment:
Just to mention so far Windows doesn't carry any leap seconds in the systems by default. I think the leap seconds will start to show up in the future.